saleor/payment/gateways/razorpay/__init__.py

```diff
@@ -0,0 +1,153 @@
+import logging
+import uuid
+from decimal import Decimal
+from typing import Dict, Tuple
+
+from django.utils.translation import pgettext_lazy
+
+import razorpay
+from razorpay import errors
+
+from ... import TransactionKind
+from ...models import Payment, Transaction
+from ...utils import create_transaction
+from .forms import RazorPaymentForm
+
+# Define the existing error messages as lazy `pgettext`.
+ERROR_MSG_ORDER_NOT_CHARGED = pgettext_lazy(
+    'Razorpay payment error', 'Order was not charged.')
+ERROR_MSG_INVALID_REQUEST = pgettext_lazy(
+    'Razorpay payment error', 'The payment data was invalid.')
+ERROR_MSG_SERVER_ERROR = pgettext_lazy(
+    'Razorpay payment error', 'The order couldn\'t be proceeded.')
+
+# Define what are the razorpay exceptions,
+# as the razorpay provider doesn't define a base exception as of now.
+RAZORPAY_EXCEPTIONS = (
+    errors.BadRequestError, errors.GatewayError, errors.ServerError)
+
+# Get the logger for this file, it will allow us to log
+# error responses from razorpay.
+logger = logging.getLogger(__name__)
+
+
+def _generate_transaction(
+        payment: Payment,
+        kind: str,
+        amount: Decimal,
+        *,
+        id='',
+        is_success=True,
+        **data) -> Transaction:
+    """Creates a transaction from a Razorpay's success payload
+    or from passed data."""
+    transaction = create_transaction(
+        payment=payment,
+        kind=kind,
+        amount=amount,
+        currency=data.pop('currency', payment.currency),
+        gateway_response=data,
+        token=id,
+        is_success=is_success)
+    return transaction
+
+
+def get_error_message_from_razorpay_error(exc: BaseException):
+    """Convert a error razorpay error to a user friendly error message
+    and log the exception to stderr."""
+    logger.exception(exc)
+    if isinstance(exc, errors.BadRequestError):
+        return ERROR_MSG_INVALID_REQUEST
+    else:
+        return ERROR_MSG_SERVER_ERROR
+
+
+def get_error_response(amount: Decimal) -> dict:
+    """Create a place holder response for invalid/ failed requests
+    for generated a failed transaction object."""
+    return {'is_success': False, 'amount': amount}
+
+
+def get_amount_for_razorpay(amount: Decimal) -> int:
+    """Convert a decimal amount to int, by multiplying the value by 100."""
+    return int(amount * 100)
```

Comment:
So actually I'm worried about this function. 

How can we ensure that amount will always have 2 decimal places? 
If the store's admin provides the wrong amount for a product, then his customers will be charged tens times more than they should be.

I think there are currencies which have more than 2 decimal or have no at all.
On the other hand, shifting the numbers will not work as well.
Eg.
"1" -> 1
"1.01" -> 101
"1.000001" -> ...

Does RazorPay allow other currencies?

Reply:
I see what you mean.

Razorpay always gives us a value in integer that has two decimal places.

It is this instruction that will tell razorpay how much we are charging (it still has the `int() * 100`):
https://github.com/mirumee/saleor/blob/97bb5ebfd924c9d6068e78547a16aef815915c15/saleor/payment/gateways/razorpay/forms.py#L35

So, it will see the converted, let's guess, USD, so in the settings, we have `DEFAULT_DECIMAL_PLACES = get_currency_fraction('USD')  # -> 2`. But will fail for a store having `BTC` as default currency.

I will fix that.

For your question: RazorPay only support the INR (Indian) currency.

Reply:
Would you be okay with doing like so:
```python
int(Decimal('3.14159').quantize(Decimal('1.01')) * 100)  # -> 3.15
```

Reply:
I'd think of something like 
```
# One indian rupee consists of 100 paisa coins
PAISA = Decimal('0.01')

def get_amount_as_paisa(money):
    """Razorpay expects Indian Rupees to be sent in their subdivision - paisa.
    """
    if money.currency != `Indian rupee`:
        # handle error
    amount_in_paisa = money.amount // PAISA
    return amount_in_paisa
```

Reply:
But, in my mind during the writing of this code, I assume that they *may* support other currencies. e.g. USD is almost supported but they deny you from paying. If you attempt to pay in EUR, they will show the error "Sorry, your currency is not supported.".

Should we really check that? I feel like it is not needed in our job, and then, if they happen to support USD, EUR for example, that would require us to change our internal gateway logic.

Reply:
I assume that currency check is up to Razorpay's gateway settings?

Maybe in such case, we could try to get the number of decimal places for currency used to make it more error-proof? And then handle the amount accordingly...

Reply:
Do their docs specify that the amount it so be given in the currency's lowest denomination or do they say that the amount is to be given in 1/100s of the main denomination? If it's the latter then I'd assume then to expect to multiply by 100 even when dealing with Japenese Yen that does not have a <1 denomination.

Reply:
The docs doesn't tell anything else than
> The amount of payment in paisa. e.g. If amount = 100, it means Rs 1.
https://docs.razorpay.com/docs/return-objects#payment-entity

Reply:
In that case I propose that we explicitly check that the currency is INR and add a comment that we need to revisit this logic if other currencies are going to be supported.

Reply:
Alright, I will do that whenever I have a few minutes free. Thank you for your input, @patrys !

Reply:
I just pushed 57338e1de14bfccd3ae532897d0b05dac0ceaa71, let me know what you think about doing this way.

Then, we could refactor later depending on how we will implement #3250.

Reply:
I just still need to write a test for `check_payment_supported`, I totally forgot about this, until going off.

![2m0op1](https://user-images.githubusercontent.com/6186720/48195499-1ab8af00-e350-11e8-9163-b77880fe3876.jpg)